src/s_net.c

```diff
@@ -0,0 +1,261 @@
+}
+
+void addrinfo_print_list(struct addrinfo **ailist)
+{
+    const struct addrinfo *ai;
+    char addrstr[INET6_ADDRSTRLEN];
+    for (ai = (*ailist); ai != NULL; ai = ai->ai_next)
+    {
+        void *addr;
+        char *ipver;
+        if (ai->ai_family == AF_INET6)
+        {
+            struct sockaddr_in6 *sa6 = (struct sockaddr_in6 *)ai->ai_addr;
+            addr = &(sa6->sin6_addr);
+            ipver = "IPv6";
+        }
+        else if (ai->ai_family == AF_INET)
+        {
+            struct sockaddr_in *sa4 = (struct sockaddr_in *)ai->ai_addr;
+            addr = &(sa4->sin_addr);
+            ipver = "IPv4";
+        }
+        else continue;
+        INET_NTOP(ai->ai_family, addr, addrstr, INET6_ADDRSTRLEN);
+        printf("%s %s\n", ipver, addrstr);
+    }
+}
+
+const char* sockaddr_get_addrstr(const struct sockaddr *sa, char *addrstr,
+    int addrstrlen)
+{
+	const void *addr;
+	addrstr[0] = '\0';
+	if (sa->sa_family == AF_INET6)
+    {
+        struct sockaddr_in6 *sa6 = (struct sockaddr_in6 *)sa;
+        addr = (const void *)&sa6->sin6_addr.s6_addr;
+    }
+    else if (sa->sa_family == AF_INET)
+    {
+        struct sockaddr_in *sa4 = (struct sockaddr_in *)sa;
+        addr = (const void *)&sa4->sin_addr.s_addr;
+    }
+    else return NULL;
+	return INET_NTOP(sa->sa_family, addr, addrstr, addrstrlen);
+}
+
+unsigned int sockaddr_get_port(const struct sockaddr *sa)
+{
+	if (sa->sa_family == AF_INET6)
+    {
+        struct sockaddr_in6 *sa6 = (struct sockaddr_in6 *)sa;
+        return ntohs(sa6->sin6_port);
+    }
+    else if (sa->sa_family == AF_INET)
+    {
+        struct sockaddr_in *sa4 = (struct sockaddr_in *)sa;
+        return ntohs(sa4->sin_port);
+    }
+    else return 0;
+}
+
+int sockaddr_is_multicast(const struct sockaddr *sa)
+{
+    if (sa->sa_family == AF_INET6)
+    {
+        struct sockaddr_in6 *sa6 = (struct sockaddr_in6 *)sa;
+        return (sa6->sin6_addr.s6_addr[0] == 0xFF);
+    }
+    else if (sa->sa_family == AF_INET)
+    {
+        struct sockaddr_in *sa4 = (struct sockaddr_in *)sa;
+        return ((ntohl(sa4->sin_addr.s_addr) & 0xF0000000) == 0xE0000000);
+    }
+    return 0;
+}
+
+int socket_connect(int socket, const struct sockaddr *addr,
+                   socklen_t addrlen, float timeout)
+{
+    int status;
+    struct timeval timeoutval;
+    fd_set writefds;
+
+    // set nonblocking and connect
+    socket_set_nonblocking(socket, 1);
+    status = connect(socket, addr, addrlen);
+    if (status < 0 && socket_errno() != EINPROGRESS)
+        return status;
+
+    // block with select using timeout
+    if (timeout < 0) timeout = 0;
+    FD_ZERO(&writefds);
+    FD_SET(socket, &writefds); // socket is connected when writable
+    timeoutval.tv_sec = (int)timeout;
+    timeoutval.tv_usec = (timeout - timeoutval.tv_sec) * 1000000;
+    status = select(socket+1, NULL, &writefds, NULL, &timeoutval);
+    if (status < 0 || !FD_ISSET(socket, &writefds))
+        return status;
```

Comment:
This doesn't work. If `select` times out, the FD_SET is empty but select returns 0, so by returning `status` you accidentally return success.

In the client code, we only care if the function succeed and the actual error code is obtained by `socket_errorno()`. So why not just return a "boolean" here? 1 = success, 0 = fail.

Reply:
I liked how in the previous version you did `error("connection timed out");` in case `writefds` was empty.

Also note that on Windows `sys_sockerr` only handles WSA errors, so maybe we should print an error message here if `select` fails?

Reply:
I went ahead and fixed this in b5ea83b